torch/onnx/utils.py

```diff
@@ -121,6 +121,81 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
     if isinstance(args, torch.autograd.Variable):
         args = (args, )
 
+    torch_out = None
+
+    if isinstance(model, torch.jit.ScriptModule):
+        proto, export_map = _export_script(model, args, f, export_params, verbose, training,
+                                           input_names, output_names, aten, export_type)
+    else:
+        proto, export_map, torch_out = _export_trace(model, args, f, export_params, verbose, training,
+                                                     input_names, output_names, aten, export_type)
+
+    if export_type == ExportTypes.PROTOBUF_FILE:
+        assert(len(export_map) == 0)
+        torch.serialization._with_file_like(f, "wb", lambda f: f.write(proto))
+    elif export_type in [ExportTypes.ZIP_ARCHIVE, ExportTypes.COMPRESSED_ZIP_ARCHIVE]:
+        import zipfile
+        compression = zipfile.ZIP_DEFLATED \
+            if export_type == ExportTypes.COMPRESSED_ZIP_ARCHIVE \
+            else zipfile.ZIP_STORED
+        with zipfile.ZipFile(f, 'w', compression=compression) as z:
+            z.writestr(ONNX_ARCHIVE_MODEL_PROTO_NAME, proto)
+            for k, v in export_map.items():
+                z.writestr(k, v)
+    elif export_type == ExportTypes.DIRECTORY:
+        import os
+        if os.path.exists(f):
+            assert(os.path.isdir(f))
+        else:
+            os.makedirs(f)
+
+        model_proto_file = os.path.join(f, ONNX_ARCHIVE_MODEL_PROTO_NAME)
+        torch.serialization._with_file_like(
+            model_proto_file, "wb", lambda f: f.write(proto))
+
+        for k, v in export_map.items():
+            weight_proto_file = os.path.join(f, k)
+            torch.serialization._with_file_like(
+                weight_proto_file, "wb", lambda f: f.write(v))
+    else:
+        raise RuntimeError('Unknown export type')
+
+    return torch_out
+
+
+def _export_script(model, args, f, export_params=True, verbose=False, training=False,
+                   input_names=None, output_names=None, aten=False, export_type=ExportTypes.PROTOBUF_FILE):
+    try:
+        graph = model.__getattr__('forward').graph()
+    except AttributeError:
+        # TODO: just trace it
+        raise RuntimeError('\'forward\' method must be a script method')
+
+    # Need to at least associate input shapes with inputs. If we propagate the
+    # rest of the shapes that's an added benefit
+    graph.propagate_shapes(args, False)
+    graph = _optimize_graph(graph, aten)
+
+    _set_input_and_output_names(graph, input_names, output_names)
+
+    if verbose:
+        print(graph)
```

Comment:
The code in `_export_trace` is now weirdly asymmetric with the code here. Is there any reason `_export_trace` couldn't grab the graph out of the trace and then operate on that exclusively?

Reply:
Or, put differently, what's the difference between `graph.export` and `trace.export`?

Reply:
@ezyang I've removed `trace.export` and re-organized this function a bit